chains/ethereum/forwarder_client.go

```diff
@@ -0,0 +1,219 @@
+
+// Keeps track of forwarder info include the state of the current
+// nonces
+type ForwarderClient struct {
+	client             *ethclient.Client
+	forwarderAddress   common.Address
+	forwarderAbi       abi.ABI
+	forwarderNonceLock sync.Mutex
+	forwarderNonce     *big.Int
+	fromAddress        common.Address
+}
+
+func NewForwarderClient(
+	client *ethclient.Client,
+	forwarderAddress common.Address,
+	fromAddress common.Address,
+) (*ForwarderClient, error) {
+	forwarderAbi, err := abi.JSON(strings.NewReader(ForwarderAbi))
+	if err != nil {
+		return nil, err
+	}
+
+	return &ForwarderClient{
+		client:           client,
+		forwarderAddress: forwarderAddress,
+		fromAddress:      fromAddress,
+		forwarderAbi:     forwarderAbi,
+	}, nil
+}
+
+func (c *ForwarderClient) GetOnChainNonce() (*big.Int, error) {
+	packed, err := c.forwarderAbi.Pack("getNonce", c.fromAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	data := eth.CallMsg{
+		To:    &c.forwarderAddress,
+		Data:  packed,
+		From:  c.fromAddress,
+		Gas:   200000,
+		Value: big.NewInt(0),
+	}
+
+	callOpts := &bind.CallOpts{From: c.fromAddress, Context: context.Background()}
+	res, err := c.client.CallContract(callOpts.Context, data, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := c.forwarderAbi.Unpack("getNonce", res)
+	if err != nil {
+		return nil, err
+	}
+
+	nonce := p[0].(*big.Int)
+
+	return nonce, nil
+}
+
+// Locks nonce access and returns the next nonce
+func (c *ForwarderClient) LockAndNextNonce() (*big.Int, error) {
+	c.forwarderNonceLock.Lock()
+
+	if c.forwarderNonce == nil {
+		return c.GetOnChainNonce()
+	} else {
+		return c.forwarderNonce.Add(c.forwarderNonce, big.NewInt(1)), nil
+	}
+}
+
+// Unlocks nonce access and sets the provided nonce
+// If transaction usage of a nonce failes the nonce shouldnt be
+// updated upon unlock, instead nil should be supplied
+func (c *ForwarderClient) UnlockAndSetNonce(nonce *big.Int) {
+	if nonce != nil {
+		c.forwarderNonce = nonce
+	}
+	c.forwarderNonceLock.Unlock()
+}
```

Comment:
Purely a style consideration, but maybe it would be better to have a separate `UnlockNonce` method with no arguments, rather than overloading this one.

Reply:
I tried out the suggestion but decided to keep it the way it was originally. I didnt like the idea of two unlock methods, and I want the programmer to think each time whether they want the nonce to be updated. Rather than naively unlocking - especially since there's another Unlock() function on the client that does update the nonce.